Algorithms/Knapsack/DynamicProgrammingKnapsackSolver.cs

```diff
@@ -0,0 +1,93 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+
+namespace Algorithms.Knapsack
+{
+    /// <summary>
+    /// Dynamic Programming Knapsack solver.
+    /// </summary>
+    /// <typeparam name="T">Type of items in knapsack.</typeparam>
+    public class DynamicProgrammingKnapsackSolver<T> : IKnapsackSolver<T>
+    {
+        /// <summary>
+        /// Returns the knapsack containing the items that
+        /// maximize value while not exceeding weight capacity.
+        /// </summary>
+        /// <param name="items">The list of items from which we select ones to be in the knapsack.</param>
+        /// <param name="capacity">The maximum weight capacity of the knapsack
+        /// to be filled. Only integer values of this capacity are tried. If
+        /// a greater resolution is needed, trying multiplying the
+        /// weights/capacity by a factor of 10.</param>
+        /// <param name="weightSelector">A function that returns the value of the specified item
+        /// from the <paramref name="items">items</paramref> list.</param>
+        /// <param name="valueSelector">A function that returns the weight of the specified item
+        /// from the <paramref name="items">items</paramref> list.</param>
+        /// <returns>The array of items that provides the maximum value of the
+        /// knapsack without exceeding the specified weight <paramref name="capacity">capacity</paramref>.</returns>
+        public T[] Solve(T[] items, double capacity, Func<T, double> weightSelector, Func<T, double> valueSelector)
+        {
+            int maxCapacity = Convert.ToInt32(Math.Floor(capacity));
+            Func<T, int> intWeightSelector = x => Convert.ToInt32(Math.Ceiling(weightSelector(x)));
+            var memoTable = Memoize(items, intWeightSelector, valueSelector, maxCapacity);
+            return GetOptimalItems(items, intWeightSelector, memoTable, maxCapacity);
+        }
+
+        private static T[] GetOptimalItems(T[] items, Func<T, int> weightSelector, double[,] memoTable, int capacity)
+        {
+            int currentCapacity = capacity;
+
+            var result = new List<T>();
+            for (int i = items.Count() - 1; i >= 0; i--)
+            {
+                if (memoTable[i + 1, currentCapacity] > memoTable[i, currentCapacity])
+                {
+                    var item = items[i];
+                    result.Add(item);
+                    currentCapacity -= weightSelector(item);
+                }
+            }
+
+            result.Reverse(); // we added items back to front

```

Comment:
The order of items in result doesn't matter as originally it should be a set, not an array. It's probably a design flaw that interface uses array and I will think about refining IKnapsackSover interface.

Reply:
I have found order to be important in the past only when dealing with different items with the same weight and value. At that point, the order can determine which is selected and which isn't. That being said, it would probably be important to either: 1) Make which item is selected under those circumstances a known or even configurable feature of the solver, or 2) Make the data structure on the interface order inspecific, as you suggest, and advertise that the result is 1 possible solution, not all solutions and not any specific solution.   The 1st is more useful from a business perspective, the 2nd probably does a better job of teaching the algorithm, which is my understanding of the goal here.

Knowing that you are considering changing the interface, I am going to go back to my implementation and make the tests resilient to that change. That is, instead of depending on the ordered nature of the array object to determine if the tests pass, I will order the results specifically in the test and validate them that way.  However, I will leave the Reverse() method call in place for right now while the data structure respects order. If the interface changes in the future, it will be easy to remove this and add documentation saying that the items should be treated as a set.

The change I would really like to see on the interface is to make the weights, including capacity, integral only. This would avoid the conversion in the Solve method and make it more explicit that the algorithm only supports integral weights. The values of course, could remain as doubles.  I explored making this change myself but it would also require a change to the ShannonFanoCompressor which depends on this interface and which I don't know enough about to be able to evaluate the impact.

Reply:
I see your point and you're totally correct about goals of this repo. Regarding changing the interface, I don't think it's a good idea because, in general, it's not specified that weights are integers, they can be real numbers as well. Where did you find that weights have to be integers in this algorithm? I've seen it on english wikipedia with no explanation and this precondition isn't present on russian wikipedia page explaining the same algorithm. As I haven't found the reason why weights can't be floating-point numbers, I think that the problem stems from the fact that weights are used as indices in array. Keeping that in mind, I changed array to dictionary so that it works like associative array. What do you think about it? All tests passed and it seems to me that it should work fine not taking into account some weird situations where floating-point arithmetic causes loss of precision.

Reply:
I've found an issue with real numbers, will try to fix it and write you back. The problem is in the way tabulation is done, I think I can fix it, but at this point it will be different algorithm because of my changes. I propose to remove interface from the class definition for now and work with integers as algorithm implies. If the code is ready, I can merge now. Will think on how to glue it all together later.

Reply:
Yes, that is correct. The tabulation is a discreet operation and thus needs to be integral for the DP algorithm to function properly. Non integers should be handled using multiplication as discussed earlier to make all weights integer values.  As best I can tell, everything looks good and the tests do pass so go ahead and merge whenever you have the opportunity and I'll validate it again once all is done. If I can assist in any way please let me know.

BTW: Thanks for doing this...